api/envoy/extensions/filters/http/credential_injector/v3/credential_injector.proto

```diff
@@ -0,0 +1,106 @@
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.credential_injector.v3";
+option java_outer_classname = "CredentialInjectorProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/credential_injector/v3;credential_injectorv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: Credential Injector]
+// [#not-implemented-hide:]
+// Credential Injector :ref:`configuration overview <config_http_filters_credential_injector>`.
+// [#extension: envoy.filters.http.credential_injector]
+
+// Filter config.
+message CredentialInjector {
+  // Credential to inject into proxied requests.
+  Credential credential = 1 [(validate.rules).message = {required: true}];
+}
+
+// Credential to inject into proxied requests.
+//
+message Credential {
+  oneof credential_type {
+    option (validate.required) = true;
+
+    // Basic Auth Credential to inject.
+    BasicAuthCredential basic = 1;
+
+    // Bearer Token Credential to inject.
+    transport_sockets.tls.v3.SdsSecretConfig bearer = 2;
+
+    // OAuth2 Credential to inject.
+    OAuth2Credential oauth2 = 3;
+  }
+}
+
+// Basic Auth Credential to inject into proxied requests.
+//
+// RFC https://www.rfc-editor.org/rfc/rfc7617
+//
+message BasicAuthCredential {
+  // User name.
+  string username = 1 [(validate.rules).string = {min_len: 1}];
+
+  // User password.
+  transport_sockets.tls.v3.SdsSecretConfig password = 2
+      [(validate.rules).message = {required: true}];
+}
+
+// OAuth2 Credential to inject into proxied requests.
+//
+message OAuth2Credential {
+  enum AuthType {
+    // The ``client_id`` and ``client_secret`` will be sent in the URL encoded request body.
+    // This type should only be used when Auth server does not support Basic authentication.
+    URL_ENCODED_BODY = 0;
+
+    // The ``client_id`` and ``client_secret`` will be sent using HTTP Basic authentication scheme.
+    BASIC_AUTH = 1;
+  }
+
+  // Credentials to authenticate client to the authorization server.
+  // OAuth RFC https://www.rfc-editor.org/rfc/rfc6749#section-2.3
+  message ClientCredentials {
+    // Client ID.
+    // OAuth RFC https://www.rfc-editor.org/rfc/rfc6749#section-2.3.1
+    string client_id = 1 [(validate.rules).string = {min_len: 1}];
+
+    // Client secret.
+    // OAuth RFC https://www.rfc-editor.org/rfc/rfc6749#section-2.3.1
+    transport_sockets.tls.v3.SdsSecretConfig client_secret = 2
+        [(validate.rules).message = {required: true}];
+
+    // OAuth RFC https://www.rfc-editor.org/rfc/rfc6749#section-2.3.1
+    // The method to use when sending credentials to the authorization server. If this setting is
+    // not specified, the value defaults to BASIC.
```

Comment:
I think enum can't be optional? So that would mean it's URL_ENCODED_BODY is not specified?

Reply:
We should make BASIC_AUTH as the first element in the enum because according to the oauth spec, the Basic authentication scheme is used by default.